### ISSUE TITLE
Bring in dependency banning and/or restricting rule(s)

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [JoshuaKGoldberg/eslint-plugin-package-json#54]
Splitting out from #42: `npm-package-json-lint` has a family of rules that can ban variants of package dependencies:

* https://npmpackagejsonlint.org/docs/rules/dependencies/no-restricted-dependencies
* https://npmpackagejsonlint.org/docs/rules/dependencies/no-restricted-devDependencies
* https://npmpackagejsonlint.org/docs/rules/dependencies/no-restricted-pre-release-dependencies
* https://npmpackagejsonlint.org/docs/rules/dependencies/no-restricted-pre-release-devDependencies

Additionally, other rules exist in that package and elsewhere for restricting dependency ranges:

* https://github.com/kellyselden/eslint-plugin-json-files/blob/master/docs/rules/no-branch-in-dependencies.md
* https://github.com/kellyselden/eslint-plugin-json-files/blob/master/docs/rules/restrict-ranges.md

I'd like to add in some kind of rule or rules to this package equivalent to those. Not clear yet how they'd look. Should they be one big configurable rule? Granular rules like how `npm-package-json-lint` & co have them? More thought required.

~Blocked on #40. Not a high priority issue given that this wouldn't be enabled by default in any recommended configs.~ ✅

**Comment by MorrisonCole** [comment]
I'd love to see this - definitely something missing from the eslint ecosystem at the moment!

https://eslint.org/docs/latest/rules/no-restricted-imports#rule-details might be a good reference too, since it does something similar.

Some granularity would be nice - one thing I've struggled with in large monorepos is gradually phasing out a dependency, which requires having warnings for some dependencies (i.e., those being removed), and errors for others (i.e., those already removed) in parallel.

**Comment by michaelfaith** [comment]
> https://eslint.org/docs/latest/rules/no-restricted-imports#rule-details might be a good reference too, since it does something similar.

This raises a good point about the value of a new rule for this.  If the usage sites of a restricted package is being reported using `no-restricted-imports` what additional value would a rule applied to the `package.json` for the same package provide?  Wouldn't that be reporting effectively the same thing in two different places?  I guess one area that this would cover that that rule doesn't, is dev tool packages that are only being used in package.json scripts and not imported somewhere where the restricted imports rule would cover?  I guess my main concern is having two different rules firing errors for the same issue.  

> one thing I've struggled with in large monorepos is gradually phasing out a dependency

The way I would see this going in practice is adding the package as a restricted import using `no-restricted-imports`, disabling existing usage and gradually reducing usage until it's zero, at which point knip would report the package as unused and it's removed entirely.  I.e. the place where this rule would fit in seems rather narrow, unless I'm missing something.

**Comment by MorrisonCole** [comment]
>This raises a good point about the value of a new rule for this. If the usage sites of a restricted package are being reported using no-restricted-imports what additional value would a rule applied to the package.json for the same package provide? Wouldn't that be reporting effectively the same thing in two different places

In a sense, yes, but I'd add that in a large monorepo with a lot of linter rules, reducing the number of rules actually run on source files is a priority. In general running rules on `package.json` is very cheap (i.e., performed infrequently), whereas running rules on source code is quite expensive.

> The way I would see this going in practice is adding the package as a restricted import using no-restricted-imports, disabling existing usage and gradually reducing usage until it's zero, at which point knip would report the package as unused and it's removed entirely.

That's exactly the approach we take yes!

The point I was trying to raise was that in a monorepo with many dependencies, it's often the case that you have dependencies at _differenct_ stages of removal. E.g., use dependency A = warning, use dependency B = error (fail CI). The trouble though is that since ESLint in general doesn't support using the same rule at different warn levels, it's not very easy (short of copying the rule itself) to implement.

If this rule is added, I was hoping it would come in two variants, for example (not sure if the names are right but):
* `no-restricted-dependencies` // can set to warn to start phasing a dependency out
* `no-banned-dependencies` // can set to error to prevent any phased out dependencies from being re-added

**Comment by JoshuaKGoldberg** [comment]
No, I don't think so. That plugin has an intentionally very opinionated set of rules for a specific use case. This rule is more of a general, lower-level building block for the general use case.

**Comment by lishaduck** [comment]
> No, I don't think so. That plugin has an intentionally very opinionated set of rules for a specific use case. This rule is more of a general, lower-level building block for the general use case.

Yeah, but if I was writing a shareable config, I could wire up module-replacements to be the defaults for this rule and skip a dependency? Not as convenient, sure, but this would handle that use-case?

**Comment by JoshuaKGoldberg** [comment]
I suppose you could 😄

**Comment by michaelfaith** [comment]
Coming back to this: there was never a new child issue created from the discussion above, so I'm going to go ahead and close this tracking issue out as complete.  Feel free to create a new more focused issue for anything additional that's remaining.
